### PR TITLE
0.2/148 add auth service to talk to backend

### PR DESCRIPTION
**Pull request by charlesstrange2**

### CODE REVIEW

File: fixtures/fixtures.js [hunk 1]
```diff
@@ -24,6 +30,7 @@ Fixtures.apply = function(mock) {
         console.log(data);
         return [200, data];
         });
+    mock.onPost('/login').reply(200,this.oauthUserToken);
```
Comment (katjam) [inline fixtures/fixtures.js]: space after 200,

File: src/main.js [hunk 2]
```diff
@@ -22,11 +22,11 @@ const routes = [
 { path: '/tap', component: Tap },
 { path: '/payment', component: Payment },
 { path: '/', component: Login }
```
Comment (katjam) [inline src/main.js]: indent?

File: src/pages/Login.vue [hunk 3]
```diff
@@ -29,7 +29,45 @@
...
+                auth: Store.auth
+            }
+        },
+        watch: {
+            /**
+             *  Watches this.auth to check for changes
+             */
+            auth : function(val) {
+                if (this.auth) {
+                    // This function fires but router doesn't appear to be doing anything...
```
Comment (katjam) [inline src/pages/Login.vue]: So is this broken? Take out or leave in?

File: src/services/netMgr.js [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            common: {
+                'X-Requested-With': 'XMLHttpRequest', // for laravel
+            }
+        }
+    })
+};
+
+/**
+ * A function that works out if we really are Authenticated.
+ * Rather than what the Store.auth actially claims.
```
Comment (katjam) [inline src/services/netMgr.js]: actially, eh?

File: src/services/netMgr.js [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+ * AXIOS post wrapper;
+ *
+ * @param route
+ * @param postData
+ * @param cb
+ * @param err
+ * @returns {Promise.<TResult>}
+ */
+
+// TODO: these are awefully similar functions... merege them?
```
Comment (katjam) [inline src/services/netMgr.js]: Loving the spelling here :bee:

File: src/services/netMgr.js [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+/**
+ * Default error handler, if one isn't provided to the api* ones.
+ * @param error
+ */
+
+NetMgr.logAJAXErrors = function (error) {
+    if (error.response) {
+        // non 2xx response
+        console.log(error.response);
```
Comment (katjam) [inline src/services/netMgr.js]: keeping all these logs in for now?

File: src/services/netMgr.js [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+NetMgr.isMocked = function () {
+    // TODO: make tis more effective
+    return this.mocker;
+};
+
+/**
+ * Sets up or restores a Mock Adapter
+ */
+
+NetMgr.mockOn = function () {
```
Comment (katjam) [inline src/services/netMgr.js]: :guitar:

File: src/services/netMgr.js [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    } else {
+        this.mocker = new MockAdapter(this.axiosInstance);
+        Fixtures.apply(this.mocker);
+    }
+};
+
+/**
+ * Removes and stores Mock Adapter and restores original Axios instance it hides
+ */
+
```
Comment (katjam) [inline src/services/netMgr.js]: Could fit a bus through there.

File: src/services/netMgr.js [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+        // Get the error from that request.
+        var origResp = error.response;
+
+        // Is it a 401 we havn't seen before? (and do we have an old token set)
+        if (origResp.status === 401 && !origCfg._retry && this.token) {
+            switch (origResp.data.error) {
+                case "invalid_token"    : // oAuth2 token invalid.
+                case "Unauthorized"     : // Official "you're logged out".
+                case "Unauthenticated." : // What Passport is returning for "your'e token has expired. Odd.
```
Comment (katjam) [inline src/services/netMgr.js]: spelling of your

File: src/services/netMgr.js [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+        var origCfg = error.config;
+
+        // Get the error from that request.
+        var origResp = error.response;
+
+        // Is it a 401 we havn't seen before? (and do we have an old token set)
+        if (origResp.status === 401 && !origCfg._retry && this.token) {
+            switch (origResp.data.error) {
+                case "invalid_token"    : // oAuth2 token invalid.
+                case "Unauthorized"     : // Official "you're logged out".
```
Comment (katjam) [inline src/services/netMgr.js]: and you'll get a 403 here - not a 401.

File: src/services/netMgr.js [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    function (error) {
+
+        // Get the original request configuration.
+        var origCfg = error.config;
+
+        // Get the error from that request.
+        var origResp = error.response;
+
+        // Is it a 401 we havn't seen before? (and do we have an old token set)
+        if (origResp.status === 401 && !origCfg._retry && this.token) {
```
Comment (katjam) [inline src/services/netMgr.js:172]: API will send Unauthenticated - if user is not logged in.

API will send Unauthorized - if user is logged in but is not permitted to see the resource.

File: src/store.js [hunk 12]
```diff
@@ -1,61 +1,115 @@
...
 };
 
-store.getRecVouchers = function() {
+/**
+ * Called from vue componenets, proxies logon process for them.
+ * @param userApiCreds
+ * @param success
+ * @param failure
+ */
+
```
Comment (katjam) [inline src/store.js]: Another bus lane?

File: src/store.js [hunk 13]
```diff
@@ -1,61 +1,115 @@
...
-        newVouchers.sort(function(b,a) {
-            return new Date(a.updated_at).getTime() - new Date(b.updated_at).getTime()
-        });
-        newVouchers = response.data.map(function(v) {
-            return v.code;
-        });
-        // this is a callback. scoping of "this" gets broken, explicitly using "store".
-        store.mergeRecVouchers(newVouchers);
-    });
+    this.netMgr.apiGet('/traders/' + this.user.id + '/vouchers',
```
Comment (katjam) [inline src/store.js]: Was this supposed to include the status= filter param as well?

File: src/store.js [hunk 14]
```diff
@@ -1,61 +1,115 @@
...
+        }.bind(this));
     return true;
 };
 
-store.mergeRecVouchers = function(replacements) {
+/**
+ * Vue's observation of arrays is tricky. This replaces the an array.
+ * @param replacements
+ */
+
```
Comment (katjam) [inline src/store.js]: extra line

File: src/store.js [hunk 15]
```diff
@@ -1,61 +1,115 @@
...
+    this.vouchers.splice(0, this.vouchers.length);
 };
 
-store.postVouchers = function() {
+
+/**
+ * Post vouchers to api.
+ * @returns {boolean}
+ */
+
```
Comment (katjam) [inline src/store.js]: I am starting to think you do it on purpose :(

**Review by katjam — COMMENTED** [review]
Pretty good... have a look at my comments and let me know.

File: src/pages/Login.vue (hunk 3 above)
Comment (charlesstrange2) [reply src/pages/Login.vue]: I'll take it out - Router wiring is part of the next card, is suppose.

File: src/services/netMgr.js (hunk 6 above)
Comment (charlesstrange2) [reply src/services/netMgr.js]: Yes; we're still dev'ing and that's a default handler for unknown oddities.

File: src/services/netMgr.js (hunk 8 above)
Comment (charlesstrange2) [reply src/services/netMgr.js]: I don't follow? :bus: ?

File: src/services/netMgr.js (hunk 11 above)
Comment (charlesstrange2) [reply src/services/netMgr.js:172]: Ah! right, will comment "Unauthorized" out and change description.
Any idea why "Unauthenticated." has that full stop?

File: src/services/netMgr.js (hunk 10 above)
Comment (charlesstrange2) [reply src/services/netMgr.js]: I thought i was getting 401s in testing. Will check.

File: src/store.js (hunk 13 above)
Comment (charlesstrange2) [reply src/store.js]: It was the function for the old "recorded" list. It might well benefit from that now, will investigate.

**Comment by charlesstrange2** [comment]
Is it not ok to put lines in?

File: src/store.js [hunk 16]
```diff
@@ -1,73 +1,115 @@
...
+store.postVouchers = function () {
     if (!navigator.onLine) {
         return false;
     }
     var postData = {
-        'trader_id' : this.trader.id,
-        'user_id'   : this.user.id,
-        'vouchers'  : this.vouchers
+        'trader_id': this.trader.id,
+        'user_id': this.user.id,
```
Comment (katjam) [inline src/store.js]: remove user_id

**Review by katjam — APPROVED** [review]
:shipit: :jamaica:

**Pull request merged**